src/private/runApi.ts

```diff
@@ -0,0 +1,145 @@
+import type { ObjWithStore } from '../useStore'
+
+type RequestType =
+  | { kind: 'fetch'; piece: CroodsState }
+  | { kind: 'destroy' }
+  | { kind: 'save'; options: SaveOptions }
+
+type Request = {
+  actions: ObjWithStore<InternalActions>
+  options: ActionOptions
+  requestType: RequestType
+}
+
+const getMethod = (options: ActionOptions, type: RequestType): HttpMethod => {
+  if (typeof options.method === 'string') return options.method
+  switch (type.kind) {
+    case 'fetch':
+      return 'GET'
+    case 'destroy':
+      return 'DELETE'
+    case 'save':
+      return options.id ? 'PUT' : 'POST'
+  }
+}
+
+const getHeaders = async (options: ActionOptions): Promise<HeadersObj> => {
+  const defaultHeaders: HeadersObj = {
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+  }
+  return typeof options.headers === 'function'
+    ? options.headers(defaultHeaders)
+    : { ...defaultHeaders, ...options.headers }
+}
+
+const getUrl = (options: ActionOptions): string => {
+  const path = buildUrl(options)
+  const queryString = buildQueryString(options)
+  return joinWith('?', path, queryString)
+}
+
+const shouldUseCache = ({ options, requestType }: Request): boolean => {
+  return Boolean(
+    requestType.kind === 'fetch' &&
+      Boolean(requestType.piece?.list?.length) &&
+      options.cache,
+  )
+}
+
+const requestAction = ({ actions, options, requestType }: Request): boolean => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getRequest(options)
+    case 'save':
+      return actions.saveRequest(options)
+    case 'destroy':
+      return actions.destroyRequest(options)
+  }
+}
+
+const successAction = <T = any>(
+  { actions, options, requestType }: Request,
+  result: T,
+): T => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getSuccess(options, result)
+    case 'save':
+      return actions.saveSuccess(options, result)
+    case 'destroy':
+      return actions.destroySuccess(options)
+  }
+}
+
+const failureAction = (
+  { actions, options, requestType }: Request,
+  error: string,
+): false => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getFail(options, error)
+    case 'save':
+      return actions.saveFail(options, error)
+    case 'destroy':
+      return actions.destroyFail(options, error)
+  }
+}
+
+async function runApi<T>(
+  request: Request,
+): Promise<(body?: ReqBody) => Promise<T>> {
+  const { actions, options, requestType } = request
+  const url = getUrl(options)
+  const headers = await getHeaders(options)
+  const method = getMethod(options, requestType)
+  return async (body?: ReqBody) => {
+    const config = { method, headers } as RequestInit
+    if (body) {
+      const parseParams = createHumps(options.paramsParser || snakeCase)
+      const data = parseParams(omit(body, 'id'))
+      config.body = JSON.stringify(data)
+    }
+    if (requestType.kind === 'fetch' && shouldUseCache(request)) {
```

Comment:
perhaps we could use this part of the code to inject the mock? Or perhaps unify with the fetch call?
I see similarities between mocking, fetching from cache and fetching from network:

* mock : get value from config -> send back as success
* cache : get value from global state -> send back as success
* fetch : get value from api endpoint -> send back as success

Reply:
I want to show you something I've done and didn't commit which uses the mock instead of the window.fetch call below, so it goes through every step of the flow and only replaces the request itself, it seems more robust IMO, I'll show you in a minute

Reply:
https://github.com/SeasonedSoftware/croods/commit/8cfa53587e1da44cc963e9d1e7f57d75af9a2184